codegen/module_system.go

```diff
@@ -65,44 +63,104 @@ type StructMeta struct {
-	EndpointReqHeaders     map[string]string      // Json blob
-	EndpointReqHeaderKeys  []string               // To keep in canonical order
-	EndpointResponse       map[string]interface{} // Json blob
-	EndpointResponseString string
-	EndpointResHeaders     map[string]string // Json blob
-	EndpointResHeaderKeys  []string          // To keep in canonical order
-
-	ClientStubs []ClientStub
-
-	TestServiceName string // The service module that mounts the endpoint
+// FixtureBlob is map[string]interface{} that implements default string
+// used for headers and (http | tchannel) request/response
+type FixtureBlob map[string]interface{}
+
+// String convert FixtureBlob to string
+func (fb *FixtureBlob) String() string {
+	str, err := json.Marshal(fb)
+	if err != nil {
+		panic(err)
+	}
+	return string(str)
+}
+
+// BodyType can either be `json` or `string`
+type BodyType string
+
+// ProtocalType can either be `http` or `tchannel`
+type ProtocalType string
+
+// FixtureBody is used to create http body in test fixtures
+type FixtureBody struct {
+	BodyType   BodyType     `json:"bodyType,omitempty"`
+	BodyString string       `json:"bodyString,omitempty"` // set BodyString if response body is string
+	BodyJSON   *FixtureBlob `json:"bodyJson,omitempty"`   // set BodyJSON if response body is object
+}
+
+// String convert FixtureBody to string
+func (fb *FixtureBody) String() string {
+	switch fb.BodyType {
+	case "string":
+		return fb.BodyString
+	case "json":
+		if fb.BodyJSON == nil {
+			panic(errors.New("invalid http body type"))
+		}
+		return fb.BodyJSON.String()
+	default:
+		panic(errors.New("invalid http body type"))
+	}
+}
+
+// FixtureHTTPResponse is test fixture for http response
+type FixtureHTTPResponse struct {
+	StatusCode int          `json:"statusCode"`
+	Body       *FixtureBody `json:"body,omitempty"`
+}
+
+// FixtureResponse is test fixture for client/endpoint response
+type FixtureResponse struct {
+	ResponseType     ProtocalType         `json:"responseType"`
+	HTTPResponse     *FixtureHTTPResponse `json:"httpResponse,omitempty"`
+	TChannelResponse *FixtureBlob         `json:"tchannelResponse,omitempty"`
+}
+
+// Body returns the string representation of FixtureResponse
+func (fr *FixtureResponse) Body() string {
+	switch fr.ResponseType {
+	case "tchannel":
+		return fr.TChannelResponse.String()
+	case "http":
+		res := ""
+		if fr.HTTPResponse.Body != nil {
+			res = fr.HTTPResponse.Body.String()
+		}
+		return res
+	default:
+		panic(errors.New("invalid response type"))
+	}
+}
+
+// EndpointTestFixture saves mocked requests/responses for an endpoint test.
+type EndpointTestFixture struct {
+	TestName           string                        `json:"testName"`
+	EndpointID         string                        `json:"endpointId"`
+	HandleID           string                        `json:"handleId"`
+	EndpointRequest    FixtureBlob                   `json:"endpointRequest"` // there's no difference between http or tchannel request
```

Comment:
This is different from your proposal.

Reply:
per our discussion previously: http request and tchannel request are the same and all can be represented with map[string]interface{}, so we reuse fixtureBlob
any suggestion here?

Reply:
I thought we agree to have a `request` type for both endpoint and client request, sth like

```
type Request struct{
     Method string
     Headers map[string]interface{}
     Body FixtureBody
}
```

The current definitions have two problems: 1. they duplicate the client request types and endpoint request types and the fields are not bundled together. 2. The request body can be of string type, same as the response body type

Reply:
headers should be aside there's already a field for it
good point on body, http request body being string I didn't see any but seems string is legit https://en.wikipedia.org/wiki/HTTP_message_body

so let's have the same idea with responseBody
```
type FixtureRequest struct {
    RequestType string
    HTTPRequest HTTPRequest
    TchannelRequest JSONBlob
}

type HTTPRequest struct {
    Method string
    Body *FixtureBody
}
```